# Проект по SQL

## Задачи

- Посчитать, сколько книг вышло после 1 января 2000 года;

- Для каждой книги посчитать количество обзоров и среднюю оценку;

- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


## Описание данных

**Таблица books**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.


**Таблица publishers**

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

## Импортируем библиотеки

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

## Устанавливаем параметры

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

## Сохраняем коннектор

In [3]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Изучим базу данных

### Напишем функцию для чтения таблиц, чтобы не писать запрос каждый раз

In [4]:
def SQL(query):
    return pd.io.sql.read_sql(query, con = engine)

### Books

In [5]:
SQL('''
    
    SELECT *,
        COUNT(*) OVER () AS table_size
    FROM books
    LIMIT 5
    
''')

,book_id,author_id,title,num_pages,publication_date,publisher_id,table_size
0,1,546,'Salem's Lot,594,2005-11-01,93,1000
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336,1000
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135,1000
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309,1000
4,5,125,1776,386,2006-07-04,268,1000


Таблица состоит из 6 столбцов и 1000 строк

### Authors

In [6]:
SQL('''
    
    SELECT *,
        COUNT(*) OVER () AS table_size
    FROM authors
    LIMIT 5
    
''')

,author_id,author,table_size
0,1,A.S. Byatt,636
1,2,Aesop/Laura Harris/Laura Gibbs,636
2,3,Agatha Christie,636
3,4,Alan Brennert,636
4,5,Alan Moore/David Lloyd,636


В таблице 2 столбца и 636 строчек

### Publishers

In [7]:
SQL('''
    
    SELECT *,
        COUNT(*) OVER () AS table_size
    FROM publishers
    LIMIT 5
    
''')

,publisher_id,publisher,table_size
0,1,Ace,340
1,2,Ace Book,340
2,3,Ace Books,340
3,4,Ace Hardcover,340
4,5,Addison Wesley Publishing Company,340


Всего 2 столбца и 340 строчек

### Ratings

In [8]:
SQL('''
    
    SELECT *,
        COUNT(*) OVER () AS table_size
    FROM ratings
    LIMIT 5
    
''')

,rating_id,book_id,username,rating,table_size
0,1,1,ryanfranco,4,6456
1,2,1,grantpatricia,2,6456
2,3,1,brandtandrea,5,6456
3,4,2,lorichen,3,6456
4,5,2,mariokeller,2,6456


Всего 4 столбца и 6456 строк

### Reviews

In [9]:
SQL('''
    
    SELECT *,
        COUNT(*) OVER () AS table_size
    FROM reviews
    LIMIT 5
    
''')

,review_id,book_id,username,text,table_size
0,1,1,brandtandrea,Mention society tell send professor analysis. ...,2793
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...,2793
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...,2793
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...,2793
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...,2793


Всего 4 столбца и 2793 строки

Изучим какие таблицы еще доступны в базе данных

In [10]:
display(pd.read_sql('''

SELECT *
FROM pg_catalog.pg_tables
WHERE schemaname != 'pg_catalog' AND
      schemaname != 'information_schema';

''', con = engine))

,schemaname,tablename,tableowner,tablespace,hasindexes,hasrules,hastriggers,rowsecurity
0,public,author,praktikum_student,None,True,False,False,False
1,public,second,praktikum_student,None,False,False,False,False
2,public,second_b,praktikum_student,None,False,False,False,False
3,public,advertisment_costs,praktikum_admin,None,True,False,False,False
4,public,authors,praktikum_admin,None,True,False,True,False
5,public,books,praktikum_admin,None,True,False,True,False
6,public,orders,praktikum_admin,None,True,False,False,False
7,public,publishers,praktikum_admin,None,True,False,True,False
8,public,ratings,praktikum_admin,None,True,False,True,False
9,public,reviews,praktikum_admin,None,True,False,True,False


Таблицы, которые нам необходимо, не единственные доступные таблицы. Теперь изучим типы столбцов в интересующих таблицах

In [11]:
display(pd.read_sql('''
SELECT 
    table_name, 
    column_name, 
    data_type, 
    is_nullable
FROM INFORMATION_SCHEMA.COLUMNS 
WHERE table_name IN ('books', 'authors', 'publishers', 'ratings', 'reviews');
''', con = engine))

,table_name,column_name,data_type,is_nullable
0,books,book_id,integer,NO
1,ratings,rating,integer,YES
2,reviews,review_id,integer,NO
3,reviews,book_id,integer,YES
4,books,num_pages,integer,YES
5,authors,author_id,integer,NO
6,books,publication_date,date,YES
7,books,publisher_id,integer,YES
8,publishers,publisher_id,integer,NO
9,books,author_id,integer,YES


Все типы данных соответствуют

## Передем к запросам

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [12]:
SQL("""
       SELECT COUNT(book_id) AS cnt_books
       FROM books
       WHERE publication_date > '2000-01-01';
    """)

,cnt_books
0,819


После 2000 года бышло 819 книг

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [13]:
SQL('''SELECT books.title, 
              books.book_id AS id,
              COUNT(DISTINCT(review_id)) as total_reviews, 
              AVG(ratings.rating) as avg_rating
                     
        FROM books
        LEFT JOIN reviews ON books.book_id = reviews.book_id
        LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY books.book_id
        ORDER BY total_reviews DESC;
          ''')

,title,id,total_reviews,avg_rating
0,Twilight (Twilight #1),948,7,3.662500
1,Water for Elephants,963,6,3.977273
2,The Glass Castle,734,6,4.206897
3,Harry Potter and the Prisoner of Azkaban (Harr...,302,6,4.414634
4,The Curious Incident of the Dog in the Night-Time,695,6,4.081081
...,...,...,...,...
995,Anne Rice's The Vampire Lestat: A Graphic Novel,83,0,3.666667
996,The Natural Way to Draw,808,0,3.000000
997,The Cat in the Hat and Other Dr. Seuss Favorites,672,0,5.000000
998,Essential Tales and Poems,221,0,4.000000


Мы получили таблицу, которая содержит название, id, количество обзоров и среднюю оценку

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [14]:
SQL('''SELECT publishers.publisher,
                     COUNT(books.book_id) as books_total
              FROM books
              LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id
              WHERE books.num_pages >= 50
              GROUP BY publishers.publisher
              ORDER BY books_total DESC
              LIMIT 1;
          ''')

,publisher,books_total
0,Penguin Books,42


Издательство Penguin Books выпустило наибольшее количество книг больше 50 страниц, их общее количество равняется 42

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [15]:
SQL('''SELECT author, 
             AVG(temp.avg_rating) AS average_rating,
             COUNT(temp.book_id) AS books_total
             FROM  (SELECT author, 
                           books.book_id, 
                           AVG(rating) AS avg_rating
                    FROM authors
                    LEFT JOIN books ON books.author_id = authors.author_id
                    LEFT JOIN ratings ON ratings.book_id = books.book_id
                    GROUP BY author, books.book_id
                    HAVING COUNT(rating) >= 50) AS temp
            GROUP BY author
            ORDER BY average_rating DESC
            LIMIT 1
           
           
        ''')

,author,average_rating,books_total
0,J.K. Rowling/Mary GrandPré,4.283844,4


Самым высокооцениваемым автором оказалась J.K. Rowling, со средним рейтингом 4.283844 и общим количеством книг в размере 4

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [16]:
SQL('''SELECT ROUND(AVG(total_reviews))
              FROM  (
                     SELECT reviews.username,
                            COUNT(reviews.review_id) AS total_reviews
                     FROM reviews
                     WHERE reviews.username IN  (
                                                 SELECT ratings.username
                                                 FROM ratings
                                                 GROUP BY ratings.username
                                                 HAVING COUNT(ratings.rating) > 48
                                                )
                     GROUP BY reviews.username
                    ) AS reviews_count;
            ''')

,round
0,24.0


Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 обзора

## Выводы

1. Всего после 1 января 2000 года вышло 819 книг
2. Twilight имеет самое большое количество обзоров и при этом средний рейтинг 3,7. Есть книги, на которые вообще нет обзоров
3. Издательство Penguin Books выпустило наибольшее количество книг больше 50 страниц, их общее количество равняется 42
4. Самым высокооцениваемым автором оказалась J.K. Rowling, со средним рейтингом 4.283844 и общим количеством книг в размере 4
5. Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 обзора